algorithms/symmetry/cosym/observers.py

```diff
@@ -0,0 +1,114 @@
+"""
+Observers for the cosym procedure.
+"""
+from collections import OrderedDict
+from dials.util.observer import Observer, singleton
+from dials.algorithms.symmetry.cosym.plots import plot_coords, plot_rij_histogram
+from dials.algorithms.clustering.observers import UnitCellAnalysisObserver
+from jinja2 import Environment, ChoiceLoader, PackageLoader
+
+
+def register_default_cosym_observers(script):
+    """Register the standard observers to the cosym script."""
+    script.cosym_analysis.register_observer(
+        event="analysed_symmetry", observer=SymmetryAnalysisObserver()
+    )
+    script.cosym_analysis.register_observer(
+        event="analysed_clusters", observer=CosymClusterAnalysisObserver()
+    )
+    script.register_observer(event="run_cosym", observer=UnitCellAnalysisObserver())
+    script.register_observer(
+        event="run_cosym", observer=CosymHTMLGenerator(), callback="make_html"
+    )
+
+
+@singleton
+class CosymHTMLGenerator(Observer):
+
+    """
+    Observer to make a html report
+    """
+
+    def make_html(self, cosym_script):
+        """Collect data from the individual observers and write the html."""
+        self.data.update(CosymClusterAnalysisObserver().make_plots())
+        self.data.update(UnitCellAnalysisObserver().make_plots())
+        self.data.update(SymmetryAnalysisObserver().make_tables())
+        filename = cosym_script.params.output.html
+        print("Writing html report to: %s" % filename)
+        loader = ChoiceLoader(
+            [
+                PackageLoader("dials", "templates"),
+                PackageLoader("dials", "static", encoding="utf-8"),
+            ]
+        )
+        env = Environment(loader=loader)
+        template = env.get_template("cosym_report.html")
+        html = template.render(
+            page_title="DIALS cosym report",
+            cosym_graphs=self.data["cosym_graphs"],
+            unit_cell_graphs=self.data["unit_cell_graphs"],
+            symmetry_analysis=self.data["symmetry_analysis"],
+        )
+        with open(filename, "wb") as f:
+            f.write(html.encode("ascii", "xmlcharrefreplace"))
+
+
+@singleton
+class CosymClusterAnalysisObserver(Observer):
+
+    """
+    Observer to record scaling model data and make model plots.
```

Comment:
I think this docstring needs updating, also the make_plots docstring in this class